gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/ReferralRequestMapper.java

```diff
@@ -0,0 +1,230 @@
+            return new Reference(PRACTITIONER_REFERENCE_PREFIX
+                + responsibleParty.getAgentRef().getId().getRoot());
+        }
+        return null;
+    }
+
+    private boolean responsiblePartyAgentRefHasIdValue(RCMRMT030101UK04ResponsibleParty3 responsibleParty) {
+        return responsibleParty != null
+            && !responsibleParty.getTypeCode().isEmpty()
+            && RESP_PARTY_TYPE_CODE.equals(responsibleParty.getTypeCode().get(0))
+            && responsibleParty.getAgentRef() != null
+            && responsibleParty.getAgentRef().getId() != null;
+    }
+
+    private Reference getRequesterAgentReference(List<RCMRMT030101UK04Participant> participantList,
+        RCMRMT030101UK04EhrComposition ehrComposition) {
+        var nonNullFlavorParticipants = participantList.stream()
+            .filter(this::isNotNullFlavourParticipant)
+            .collect(Collectors.toList());
+
+        var pprfParticipants = getParticipantReference(nonNullFlavorParticipants, PPRF_TYPE_CODE);
+        if (pprfParticipants.isPresent()) {
+            return new Reference().setReference(PRACTITIONER_REFERENCE_PREFIX + pprfParticipants.get());
+        }
+
+        var prfParticipants = getParticipantReference(nonNullFlavorParticipants, PRF_TYPE_CODE);
+        if (prfParticipants.isPresent()) {
+            return new Reference().setReference(PRACTITIONER_REFERENCE_PREFIX + prfParticipants.get());
+        }
+
+        var participant2Reference = getParticipant2Reference(ehrComposition);
+        if (participant2Reference.isPresent()) {
+            return new Reference().setReference(PRACTITIONER_REFERENCE_PREFIX + participant2Reference.get());
+        }
+
+        return null;
+    }
+
+    private Optional<String> getParticipantReference(List<RCMRMT030101UK04Participant> participantList, String typeCode) {
+        return participantList.stream()
+            .filter(participant -> hasTypeCode(participant, typeCode))
+            .filter(this::hasAgentReference)
+            .map(RCMRMT030101UK04Participant::getAgentRef)
+            .map(RCMRMT030101UK04AgentRef::getId)
+            .map(II::getRoot)
+            .findFirst();
+    }
+
+    private Optional<String> getParticipant2Reference(RCMRMT030101UK04EhrComposition ehrComposition) {
+        return ehrComposition.getParticipant2().stream()
+            .filter(participant2 -> participant2.getNullFlavor() == null)
+            .map(RCMRMT030101UK04Participant2::getAgentRef)
+            .map(RCMRMT030101UK04AgentRef::getId)
+            .map(II::getRoot)
+            .findFirst();
+    }
+
+    private boolean hasAgentReference(RCMRMT030101UK04Participant participant) {
+        return participant.getAgentRef() != null && participant.getAgentRef().getId() != null;
+    }
+
+    private boolean hasTypeCode(RCMRMT030101UK04Participant participant, String typeCode) {
+        return !participant.getTypeCode().isEmpty() && participant.getTypeCode().get(0).equals(typeCode);
```

Comment:
Do we know if given typeCode is always a first element? Maybe it would be safer to check it this way:
```suggestion
        return participant.getTypeCode()
            .stream()
            .anyMatch(typeCode::equals);
```